CGen bugs when there are continues multi assignment statement in v

<!-- Please make sure to run `v up` before reporting any issues as it may have already been fixed.
     It's also advisable to update all relevant modules using `v outdated` and `v install` -->

<!-- You can use `v doctor` to fill up the next fields -->
**V version:** 

```$ v doctor
OS: windows, Microsoft Windows 10 Home Single Language v19042 64-bit
Processor: 2 cpus, 64bit, little endian, AMD A4-9125 RADEON R3, 4 COMPUTE CORES 2C+2G
CC version: clang version 12.0.0 (https://github.com/llvm/llvm-project.git 19a5e15499282ab4ac2fa968d7f1b9d3ce97453d)

getwd: C:\Users\reguler\vtest
vmodules: C:\Users\reguler\.vmodules
vroot: E:\program\v
vexe: E:\program\v\v.exe
vexe mtime: 2021-07-22 02:44:06
is vroot writable: true
is vmodules writable: true
V full version: V 0.2.2 054bd67

Git version: git version 2.30.0.windows.2
Git vroot status: weekly.2021.29-32-g054bd674 (13 commit(s) behind V master)
.git/config present: true
thirdparty/tcc status: thirdparty-windows-amd64 9fa4d4b6
```

**OS:** 

**What did you do?**
Based a lot of discussions on [vodcha bugs](https://discord.com/channels/592103645835821068/592114487759470596/867601903986802728), there are bugs in cgen when generating output c files from func/methods contains continues multi assignment expression.
This minimal code sample reproduces the bugs 
```
module main

fn round(a int, b int, c int) (int, int, int) {
	mut ax := a 
	mut bx := b 
	mut cx := c 

	ax += bx 
	bx ^= cx 
	cx -= 1

	return ax, bx, cx 
}

fn round_assignment(mut res []int){
    // here we reproduce the bug, when first assignment commented, its run succesfully
    _ = res[3] // bound early check
    res[0], res[1], res[2] = round(res[0], res[1], res[2])
}

fn main() {
   	mut  a := []int{len:3}
	round_assignment(mut a)
	println(a)
}
```

**What did you expect to see?**
```
$ v run main.v
[0, 0, -1]
```
 
**What did you see instead?**
```
$ v run main.v
==================
C:/Users/reguler/AppData/Local/Temp/v/main.16044957081161457777.tmp.c:5778: warning: implicit declaration of function 'tcc_backtrace'
C:/Users/reguler/AppData/Local/Temp/v/main.16044957081161457777.tmp.c:10726: error: unexpected end of file

==================
(Use `v -cg` to print the entire error message)

builder error:
==================
C error. This should never happen.

If you were not working with C interop, this is a compiler bug, please report the bug using `v bug file.v`.

https://github.com/vlang/v/issues/new/choose

You can also use #help on Discord: https://discord.gg/vlang
```

--- a/vlib/v/gen/c/cgen.v
+++ b/vlib/v/gen/c/cgen.v
@@ -2224,6 +2224,7 @@ fn (mut g Gen) gen_assign_stmt(assign_stmt ast.AssignStmt) {
 	}
 	mut return_type := ast.void_type
 	is_decl := assign_stmt.op == .decl_assign
+	g.assign_op = assign_stmt.op
 	op := if is_decl { token.Kind.assign } else { assign_stmt.op }
 	right_expr := assign_stmt.right[0]
 	match right_expr {
@@ -2353,20 +2354,40 @@ fn (mut g Gen) gen_assign_stmt(assign_stmt ast.AssignStmt) {
 				}
 				g.expr(lx)
 				noscan := if is_auto_heap { g.check_noscan(return_type) } else { '' }
-				if is_opt {
-					mr_base_styp := g.base_type(return_type)
-					if is_auto_heap {
-						g.writeln(' = HEAP${noscan}($mr_base_styp, *($mr_base_styp*)${mr_var_name}.data).arg$i);')
+				if g.is_arraymap_set {
+					if is_opt {
+						mr_base_styp := g.base_type(return_type)
+						if is_auto_heap {
+							g.writeln('HEAP${noscan}($mr_base_styp, *($mr_base_styp*)${mr_var_name}.data).arg$i) });')
+						} else {
+							g.writeln('(*($mr_base_styp*)${mr_var_name}.data).arg$i });')
+						}
 					} else {
-						g.writeln(' = (*($mr_base_styp*)${mr_var_name}.data).arg$i;')
+						if is_auto_heap {
+							g.writeln('HEAP${noscan}($styp, ${mr_var_name}.arg$i) });')
+						} else {
+							g.writeln('${mr_var_name}.arg$i });')
+						}
 					}
 				} else {
-					if is_auto_heap {
-						g.writeln(' = HEAP${noscan}($styp, ${mr_var_name}.arg$i);')
+					if is_opt {
+						mr_base_styp := g.base_type(return_type)
+						if is_auto_heap {
+							g.writeln(' = HEAP${noscan}($mr_base_styp, *($mr_base_styp*)${mr_var_name}.data).arg$i);')
+						} else {
+							g.writeln(' = (*($mr_base_styp*)${mr_var_name}.data).arg$i;')
+						}
 					} else {
-						g.writeln(' = ${mr_var_name}.arg$i;')
+						if is_auto_heap {
+							g.writeln(' = HEAP${noscan}($styp, ${mr_var_name}.arg$i);')
+						} else {
+							g.writeln(' = ${mr_var_name}.arg$i;')
+						}
 					}
 				}
+			}
+			if g.is_arraymap_set {
+				g.is_arraymap_set = false
 			}
 			return
 		}

--- a/vlib/v/tests/multiple_assign_array_index_test.v
+++ b/vlib/v/tests/multiple_assign_array_index_test.v
@@ -1,0 +1,24 @@
+fn round(a int, b int, c int) (int, int, int) {
+	mut ax := a
+	mut bx := b
+	mut cx := c
+
+	ax += bx
+	bx ^= cx
+	cx -= 1
+
+	return ax, bx, cx
+}
+
+fn round_assignment(mut res []int) {
+	_ = res[2]
+	res[0], res[1], res[2] = round(res[0], res[1], res[2])
+}
+
+fn test_multiple_assign_array_index() {
+	mut a := []int{len: 3}
+	round_assignment(mut a)
+
+	println(a)
+	assert a == [0, 0, -1]
+}

